### PR TITLE
Do not fail bulk requests after a refresh failure preventing checkpoints from advancing

### LINKED ISSUE #108190: A failed after write refresh can prevent advancing the local checkpoint even when the operations were made durable by the translog
Today, when an after write refresh fails, we consider the `AsyncAfterWriteAction` as failed:
https://github.com/elastic/elasticsearch/blob/7f83189cb24bf41ddea13126385c47302ea9f9b3/server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java#L466-L471

This prevents from advancing the local checkpoints with the safely persisted seq nos:
https://github.com/elastic/elasticsearch/blob/4076aa7698cbdc5871681bd31b20477e6b295bdc/server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java#L174-L195

We should reconsider this behaviour and maybe advance the local checkpoints when the refresh failed for an unpromotable shard.

### PR DESCRIPTION
**Pull request by fcofdez**
Relates #95049
Relates #94825
Closes #108190

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java [hunk 1]
```diff
@@ -498,7 +491,7 @@ public void onResponse(Boolean forceRefresh) {
 
                     @Override
                     public void onFailure(Exception e) {
-                        refreshFailure.set(e);
+                        refreshed.set(false);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:494]: I am not entirely sure this is correct. The flag we are setting ends up in `forcedRefresh` I think. That flag is not an indication of whether the refresh succeeded. Rather it is whether the indexing operation forced a refresh for various reasons.

But in the timeout case, I suspect we could end up in having a response where the search shard is not necessarily refreshed?

Or are we failing the search shard also on a timeout failure? If so, that seems dangerous.

**Review by henningandersen — COMMENTED** [review]
I have a comment/question.

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java (hunk 1 above)
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:494]: >But in the timeout case, I suspect we could end up in having a response where the search shard is not necessarily refreshed?

In the timeout case we [fail the bulk request](https://github.com/elastic/elasticsearch/blob/7f83189cb24bf41ddea13126385c47302ea9f9b3/server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java#L466-L471) too which seems wrong to me? 

> Or are we failing the search shard also on a timeout failure? If so, that seems dangerous. But given that we apply the timeout [here](https://github.com/elastic/elasticsearch/blob/42dbc3d795c94ed3068dfd45d8b92e110c9076cc/server/src/main/java/org/elasticsearch/action/support/replication/PostWriteRefresh.java#L149) I think that is not the case.

No, we're not failing the shard after a refresh timeout.

**Comment by fcofdez** [comment]
I'm closing this since this change will change the semantics of the API during refresh failures.

**Pull request closed without merging**